Added support for multiple codes executed in a row

martinfrancois opened the pull request:
**Description:**

Apparently, a certain switch I wanted to use with the raspberry pi rf switch component turned out to need not only one, but four individual codes transmitted in sequence for the switch to turn on or off. Since I noticed there is no support for that yet in the current rpi_rf.py implementation, I decided to add it. I also made changes in the docs to explain how to use multiple codes being sequentially executed.
Unfortunately, for performance reasons I had to remove the logger entry for the check if the codes could be transmitted, but since when logging the rpi_rf in debug shows codes that are being transmitted anyways, I don't think this would really be necessary, since it is still easily possible to see if the codes have been transmitted or not.
An additional benefit to this is that there is now the option of turning on / off multiple devices at once, since individual codes of devices can be chained.

Commits:
now codes can be specified either by simply providing a single code, which will then be sent like usual, or multiple codes can be executed in a row, specified in a comma delimited format in the configuration.yaml. For example: 111111,222222,333333,444444 would mean 111111 would be sent first, followed by 222222 and 333333 and 444444.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#2018

**Example entry for `configuration.yaml` (if applicable):**
```yaml
switch:
   platform: rpi_rf
   gpio: 17
   switches:
      living_room_light:
        protocol: 5
        code_on: 654321,565874,233555,149874
        code_off: 654320,565873,233554,149873
        signal_repetitions: 15
```

**Checklist:**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L16
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L51

martinfrancois committed:
Added support for multiple codes executed in a row

now codes can be specified either by simply providing a single code, which will then be sent like usual, or multiple codes can be executed in a row, specified in a comma delimited format in the configuration.yaml. For example: 111111,222222,333333,444444 would mean 111111 would be sent first, followed by 222222 and 333333 and 444444.

homeassistant commented:
Hi @martinfrancois, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#5908).

Once you do that we will be able to review and accept this pull request.

Thanks!

martinfrancois committed:
rpi_rf: added line breaks to not exceed 79 characters per line

n8henrie commented on homeassistant/components/switch/rpi_rf.py:
```diff
@@ -27,8 +27,8 @@
 SWITCH_SCHEMA = vol.Schema({
-    vol.Required(CONF_CODE_OFF): cv.positive_int,
-    vol.Required(CONF_CODE_ON): cv.positive_int,
+    vol.Required(CONF_CODE_OFF): cv.string,
```
Looking into Voluptuous, is there any way to further validate the string to verify that it is composed entirely of comma-separated ints? Seems like this change will lose the ability to ensure no letters made it into the codes and detect that during config validation (prior to run-time). Is there perhaps a way to at least specify ranges of valid characters e.g. `[0-9,]`?

EDIT: Another option that would preserve config validation may be to have the config be a yaml array of codes.

martinfrancois replied:
Thanks for the idea! I added a validation in a new commit.

n8henrie commented on homeassistant/components/switch/rpi_rf.py:
```diff
@@ -101,22 +105,23 @@ def is_on(self):
+        """Send the code(s) with a specified pulselength."""
+        _LOGGER.info("Sending code(s): %s", code_list)
+        for code in code_list:
+            self._rfdevice.tx_code(code, protocol, pulselength)
```
Is there a reason you removed the return value? Seems like [rpi_rf has a meaningful return value](https://github.com/milaq/rpi-rf/blob/master/rpi_rf/rpi_rf.py#L90) .

martinfrancois replied:
Nevermind! To clarify, with performance reasons I meant that for some reason, when transmitting the specified codes via the rpi_rf.py script directly without going through HA, the codes would get transmitted fast enough so the switch will be triggered. At first, when I modified the HA component to be able to have support for multiple codes while preserving the return value, the codes all get executed properly, but the switch would not turn on or off.
I noticed that when I removed "bulk" to make it run faster that it would turn the switch on and off. This makes me think that the additional "bulk" of having to check the return value each time takes up so much processing power while running HA at the same time that it cannot send the codes fast enough one after each other. I think that maybe for the switch to recognize all of the codes as "one sequence" they cannot have a delay between them that is longer than a specific amount of time.

n8henrie replied:
I will have to do some local tweaking, but I think you're off base -- the return value shouldn't change anything in this regard as far as I can think. I have a few other RF based libraries I've written or messed with heavily, and the type of unreliability you describe seems usually due to the CPU being busy and running on a non-real-time OS. I've improved it significantly by [changing the scheduling policy](https://github.com/n8henrie/rf_pi#disclaimer), but this requires additional permissions that would not be feasible inside of HomeAssistant as far as I can think. Again, I think you should restore the return values -- it should not be affecting the reliability, though I will pull locally and test to verify.

I have run into [issues using multiple RF codes in an async context](https://github.com/n8henrie/homebridge-rcswitch-gpiomem/issues/3), but to be honest I never figured out what the solution was, it just fixed itself. I was wondering if it was due to multiple RF codes being sent simultaneously and thereby getting mangled, or just due to CPU congestion messing with the timing. With the way this is written, I doubt they are getting sent simultaneously.

Just as a note, I initially added `signal_repetitions` to the hass component as a poor-man's workaround for these kinds of reliability issues.

martinfrancois replied:
Good to know! I honestly know not enough about these kind of things to explain it, but I can tell you that the solution the way it is now works with my switch, which needs 4 different codes sent in a row. Things like removing the conditional and other things caused it to simply not switch on or off for whatever reason.

I also tried to mess with signal_repetitions, but I didn't really help, no matter what other code than the one you see here worked.
I find it quite weird, because if I send the codes with the rpi_rf component with the example code on the repo directly without going through HA, it's no issue at all. It works even with the response loop being checked every time and it works very quickly. It just seems that with HA for whatever reason it doesn't work even though according to the debug it sends all the codes the same way, maybe just with a slightly higher delay between them and I have no idea how I can prioritize the task more.

n8henrie commented on homeassistant/components/switch/rpi_rf.py:
```diff
@@ -101,22 +105,23 @@ def is_on(self):
     def turn_on(self):
         """Turn the switch on."""
-        if self._send_code(self._code_on, self._protocol, self._pulselength):
+        if self._send_code(self._code_on_list, self._protocol,
```
If you're going to hard-code the `return True` above in `_send_code`, do you really need to check these conditionals?

martinfrancois replied:
Quite interestingly, I thought so too. It seemed that when I ran the code without the if conditional and the hardcoded return value, the switch would not turn on. I believe (but I am no expert in Python and may be wrong) if I leave out the return value entirely, then Python may execute `_send_code` in a separate thread and continue on with the rest of the lines, like setting the state etc., which means more performance will be used at the same time. I think this approach forces Python to wait for the outcome of `_send_code` before it will continue setting the state and so on, since when I changed it to be this way, the codes have been transmitted fast enough that the switch recognized them as one sequence, while the other way around it didn't.

n8henrie reviewed (changes requested):
Perhaps it would be worthwhile to investigate a way to preserve config validation, since the codes have a predictable pattern. Also, I'm not clear why the return value was hardcoded `True`.

martinfrancois committed:
include validation for correct formatting of codes

added regex which only allows either a single number (like 1252456245) or a sequence of commas followed by another number.

martinfrancois committed:
added line breaks to not exceed 79 characters per line

martinfrancois committed:
fix for 'continuation line under-indented for visual indent'

martinfrancois committed:
another try at 'continuation line under-indented for visual indent'

andrey-git commented:
@martinfrancois Any updates on this PR?

andrey-git commented:
Actually @n8henrie too :)

andrey-git commented on homeassistant/components/switch/rpi_rf.py:
```diff
@@ -27,8 +27,10 @@
 SWITCH_SCHEMA = vol.Schema({
-    vol.Required(CONF_CODE_OFF): cv.positive_int,
-    vol.Required(CONF_CODE_ON): cv.positive_int,
+    vol.Required(CONF_CODE_OFF):
```
You can do: vol.Required(CONF_CODE_OFF): [cv.positive_int] to use yaml list

If you prefer a comma-separated list use vol.All(cv.ensure_list_csv, [cv.positive_int])

martinfrancois replied:
I think the regex solution should work as well, doesn't it? Or is there a reason for me to change it to a list?

n8henrie replied:
Either the yaml list or `vol.All(cv.ensure_list_csv, [cv.positive_int])` (thanks @andrey-git) is much cleaner and more readable than the regex, I would recommend going with one of those.

balloob replied:
To keep the code maintainable please convert it to @n8henrie suggestion

martinfrancois replied:
Thanks @balloob @n8henrie @andrey-git, I converted it now!

martinfrancois commented:
For me it looks good as is, I implemented the feedback from @n8henrie. I don't know why the coverage decreased by -0.3% however?

n8henrie commented:
Sorry, busy non-tech career, a 1 year old, and a very rural location with horrible internet.

Did you add any tests? If not, I assume coverage decreased because the total code length increased without any increase in tests. Maybe not, I'm no expert.

Also, sorry I didn't think of this before, but what advantage does this provide over using [a group](https://home-assistant.io/components/group/)? If there's not a concrete advantage, it may additional complexity without any additional functionality -- I'm currently running a fork based off this PR, comparing a single rpi-rf switch that has 2 device codes (the functionality added in this PR), and a group that has the same 2 individual switches (pre-existing functionality), the two seem to work identically. Well, not identically -- for me there is perhaps a tenth of a second greater lag using a group, which in all likelihood may increase the chances of the codes going through reliably.

Also, I've made a fork of your PR and made some tweaks that preserve the return code and eliminate the unnecessary conditional. I'm not sure why you had problems this way, it's working identically to me. I'll send you a PR so you can check it out.

Anyway, have you tried just using a `group`? If not, maybe give it a shot -- it may render this PR unnecessary, works fine for me.

andrey-git commented:
Coverage report is bogus - only rpi_rf.py has changed and it is excluded from tests.

martinfrancois commented:
@n8henrie using a group would be very uncomfortable, as I would need to click on 3 different switches just to activate one switch. I think you must've forgotten about the main reason for these changes: one of the switch types I own needs 3 consecutive codes just to switch on or off 1 switch.
I'll have a look at your PR, but I think it will most likely not work, as I already tried first to implement it preserving the return type and without the "unnecessary conditional" (see my comments) and it didn't work. I think maybe you don't own a switch which needs multiple codes executed in a sequence then?
@andrey-git thanks for telling me!

n8henrie commented:
@martinfrancois have you tried putting them in a group? It sounds like we're talking about different things -- you only have to toggle one switch to activate every device in the group, that's the point of the group. While I only tried with 2, it seems to toggle the devices one right after the other just like your code did, I would imagine it would do the same with 3 or 4. Although I do see how it would be inconvenient to have to add each code to your configuration as a different switch with this setup.

What kind of switch is this? That's definitely an interesting predicament, I'd like to check these switches out. I wonder why they designed them that way.

balloob commented on homeassistant/components/switch/rpi_rf.py:
```diff
@@ -27,8 +27,10 @@
+    vol.Required(CONF_CODE_OFF):
+        vol.All(cv.string, vol.Match(r"^[\d]+(,*[\d]*)*$")),
+    vol.Required(CONF_CODE_ON):
+        vol.All(cv.string, vol.Match(r"^[\d]+(,*[\d]*)*$")),
```
Same

balloob commented:
Coverage can be ignored as this file is ignored in coveragerc.

martinfrancois committed:
changed from regex to list for easier maintainability

martinfrancois commented:
@n8henrie Actually I tried yes, but this looks quite bad for one since it seems like there are 4 individual switches but in reality there aren't, and it uses up quite a lot of space. Also, it didn't work. I think the codes are not executed quickly enough.
Regarding the switch, it was a very cheap set of 5 I purchased about 4-5 years ago in a discount store. I was wondering too, especially because it was so cheap I wouldn't have expected it to be that sophisticated.
The switch brand and model number is "Silvercrest 91210" you might do a search on this and notice a lot of users looking for a solution to use them, because of the rolling codes. I think what might be the most interesting resource considering this would be the following implementation of switching these in the pimatic project: https://github.com/pimatic/pimatic/issues/386

Considering the pull request @n8henrie proposed, as I wrote there I tried it with my switch and it didn't switch on or off. I guess it's due to the same issues I've faced and the state of this solution is the only one that has worked so far.

andrey-git commented on homeassistant/components/switch/rpi_rf.py:
```diff
@@ -84,6 +86,10 @@ def __init__(self, hass, name, rfdevice, protocol, pulselength,
         self._pulselength = pulselength
         self._code_on = code_on
```
Now code_on is already a list - you don't need to split manually

martinfrancois replied:
Thanks for the note @andrey-git, I forgot about that. I just changed it with the latest commit.

martinfrancois committed:
removed unnecessary splitting of strings

martinfrancois commented:
Since this one has been merged - maybe it would make sense to merge this one too?
https://github.com/home-assistant/home-assistant.github.io/pull/2018

balloob commented:
Done, thanks again :)
